Implement Fibonacci and NegaFibonacci sequences

undrewb opened the pull request:
Fixes #255 
Review of colleague's PR #246

Changes proposed in this PR:

Implementation of fibonacci and negafibonacci sequences as specified in lab1 of the anz go-course.
Updated to use iteration used and I've put in a limit of +-60 entries to stop overflow errors occuring.

jonathanheaden committed:
Implement fib func

Initial draft

lexual-anz committed:
Implement fibonnaci printing with fib()

Commit:
Build Lab1 Fibonacci solution and tests

Commit:
Build Lab1 negafibonacci solution and tests

Commit:
Build Lab2 Bubble sort and bonus Insertion sort solutions and test cases

lexual-anz committed:
Implement negafibonnaci, and improve test suite

Commit:
Implement Lab1

Rename variables

Edit Test

Commit:
Implement optional improvements as per reviews and include range check

Commit:
Implement fibonacci sequence for positive integers

Uses the definition of fibonacci sequence in wikipedia.
https://en.wikipedia.org/wiki/Fibonacci_number
Outputs 0 for the 0th fibonccia entry.
The implementation uses a recursive method.
It is for lab 01 of the go-couse.
https://github.com/undrewb/go-course/tree/master/01_fib

Add overflow check

Limit fibonacci call to n<=20 so that there won't be an overflow.

Implement negafibonacci

Added a sign evaluation if starting n is less than 0.
This works per the wikipedia article about fibonacci.

Resolve issues with negafibonacci

Negafibonacci is now easier to test.
Sign only takes one parameter - assumes it is being called only by negafibonacci.

Resolve linting errors

Added table tests to remove repetitive tests

Added underflow check for -20

Address issues raised in PR #213

1. Use more efficient approach for calculating sign.
Using an array lookup is faster than using pow.
2. Use table tests for sign unit testing.
Removes repetition.

Integrate feedback from #PR213

Rework logic and test based on PR feedback

Add unit test for underflow and overflow panics

Integrate PR feedback

Use loop instead of recursion
Replace panic for overflow and underflow with error printing

Use error to report out of bounds errors

Remove error check from main to ensure 100% test coverage

Remove unused import

Remove err variable from main

n0npax committed:
Complete Lab02 - sorting

Implement BubbleSort
Implement Insertion
Implement QuickSort
Add Tests

Commit:
Complete lab03 - Letter frequency

Map letters to frequency
Return ordered list of latters

Commit:
Implement Fib and add tests

Fib uses two vars to store the fib numbers as they are incremented. Tests cover positive, negative and fib(0)

Commit:
feat(lab1): Create fibonacci number generator function

Added a function that takes an integer input and prints out the fibonacci sequence with that many iterations. This also includes the ability to print out negafibonacci.

Commit:
feat(lab1): Create fibonacci number generator function

Added a function that takes an integer input and prints out the fibonacci sequence with that many iterations. This also includes the ability to print out negafibonacci.

Commit:
test(lab1): Test main function

Test written for the main function of fibonacci, lab1

Commit:
test(lab1): Test main function

Test written for the main function of fibonacci, lab1

Commit:
test(lab1): Test fib function

Add a series of tests to ensure the fibonacci function outputs correctly if given both positive and negative integers.

Commit:
test(lab1): Test fib function

Add a series of tests to ensure the fibonacci function outputs correctly if given both positive and negative integers.

Commit:
refactor(lab1): Refactor fib function to use modulus

Based on a hint from Dan in the lecture today, I have refactored my implementation to use modulus and taken a more 'positive fibonacci first' approach. I have also added comments to help provide clarity on what each section of the code is doing.

Commit:
refactor(lab1): Refactor fib function to use modulus

Based on a hint from Dan in the lecture today, I have refactored my implementation to use modulus and taken a more 'positive fibonacci first' approach. I have also added comments to help provide clarity on what each section of the code is doing.

Commit:
Change main from fib(100) to fib(7)

Commit:
fix(typo): Fixed a typo in comment

Commit:
fix(typo): Fixed a typo in comment

Commit:
Lab3 - Implement sorted letter counting

 - Add Simple app to demonstrate sorted counts
 - Implement letters function to count runes in a string
 - Implement sortLetters function to sort rune counts

n0npax committed:
Complete lab05 - Stringer

Implement IPAddr Stringer method
Add tests

n0npax commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+func sign(n int) int {
+	var result = 1
+	if n < 0 {
+		result = int(math.Pow(-1, float64(n+1)))
```
You can optimize this to sth like:
```
[2]int{1,-1}[i%2]
```

undrewb replied:
Nice one. At first glance it feels less readable but the more I look at it the more I like it as it properly captures whats going on.

anzdaddy replied:
You can make things faster, avoiding the branch and the array lookup at the expense of readability (would definitely warrant an explanatory comment):
```go
func sign(n int) int {
	const shift = 8*unsafe.Sizeof(0) - 1
	return (n+1)<<shift&n>>(shift-1) | 1
}
```
If `sign` is only called for n < 0, you can omit `&n`.

undrewb replied:
Not sure I want to get in to gofu while I am learning!. I'll stick with the readable suggestion from n0npax for now, and put the super fast one on my 'cool stuff' to look at list.

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func sign(n int) int {
+	var result = 1
```
nit: You should normally use `:=` when initialising `int` to a constant.

undrewb replied:
Done where applicable.

n0npax commented on 01_fib/undrewb/main_test.go:
```diff
@@ -0,0 +1,124 @@
+	}
+}
+
+func TestSignPositive(t *testing.T) {
```
these tests(line 102-124) can be parametrized like tests above

undrewb replied:
By paramaterized do you mean table tests?

undrewb replied:
I'm assuming table tests because I've redone it with these and its much cleaner.

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+func sign(n int) int {
+	var result = 1
+	if n < 0 {
+		result = int(math.Pow(-1, float64(n+1)))
```
nit: You can `return` directly, rather than holding a variable till the end.

undrewb replied:
Done.

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+	}
+	if n == 0 {
+		return 0
+	}
```
Since you're assuming n >= 0 (though note my comment below) you can simplify this to:
```go
if n < 2 {
	return n
}
```

undrewb replied:
Done

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func negafibonacci(n int) int {
+	return sign(-1*(n)) * fibonacci(n)
```
`-1*(n)` = `-n`

undrewb replied:
Redundant with restucture but Ive used the same syntax elsewhere.

n0npax reviewed (changes requested):
Good work! Just 2 minor issues.

The instruction also asks to provide not more than 2 commits (https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments). Please consider rebasing.

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+	for idx := 0; idx <= lastIdx; idx++ {
+		var fibonacciN int
+		if n >= 0 {
+			fibonacciN = fibonacci((idx))
```
`((…))` unnecessary.

undrewb replied:
Fixed.

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+func fib(n int) {
+	if n > 20 || n < -20 {
+		fmt.Fprintln(out, "I dont go past twenty")
+		return
```
Better to return an `error` or `panic`.

undrewb replied:
Thanks. Wasnt sure how to do this. I'll error.

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+	fib(7)
+}
+
+func sign(n int) int {
```
nit: The name `sign` is commonly expected to be -1 for negative numbers, 0 for 0, and 1 for positive numbers. Might be good to choose a different name.

undrewb replied:
Renamed to fibonacciSign

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+	return fibonacci(n-1) + fibonacci(n-2)
+}
+
+func negafibonacci(n int) int {
```
This should be defined over n < 0. E.g.: `negafibonacci(-2)` = -1. In fact, "negafibonacci" could be seen as simply an extension of the range of the fibonacci numbers, so this function wouldn't even exist.

undrewb replied:
Ive made negafibonacci because I couldnt figure out how to loop in different directions without this kind of branch coming in somewhere.

undrewb replied:
Restructured algorithm logic based on this. Thanks.

anzdaddy commented on 01_fib/undrewb/main_test.go:
```diff
@@ -0,0 +1,124 @@
+	{4, -3},
+	{5, 5},
+	{6, -8},
+	{7, 13},
```
n should be -1, -2, …, -7.

undrewb replied:
Not for my implementation. If I were testing fib(n) sure, but negafibonacci(n) takes an abs value because we have to iterate.
Wont change.

undrewb replied:
Actually done. I reworked my implementation.

anzdaddy commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	var lastIdx = int(math.Abs(float64(n)))
+	for idx := 0; idx <= lastIdx; idx++ {
```
A top-level `if n >= 0` around two for-loops would yield a cleaner code structure.

undrewb replied:
Done

anzdaddy commented on 01_fib/undrewb/main_test.go:
```diff
@@ -0,0 +1,124 @@
+	}
+}
+
+func TestSignNegativeSix(t *testing.T) {
```
nit: I would fold these into a single test with a loop covering a range of negative even numbers and another covering odd numbers. Make sure `t.Errorf` reports the `n` that caused the failure and the value produced by `sign`.

undrewb replied:
In progress.

undrewb replied:
Done

anzdaddy commented on 01_fib/kasunfdo/main.go:
```diff
@@ -0,0 +1,36 @@
+package main
```
This file doesn't belong in this PR.

undrewb replied:
It wound up in the main repo somehow. I'll remove it.

anzdaddy commented on 01_fib/kasunfdo/main_test.go:
```diff
@@ -0,0 +1,87 @@
+package main
```
This file doesn't belong in this PR.

undrewb replied:
Removed file from PR

nicholascross committed:
Lab 4 - Implement numeronyms

- Implement function that returns a slice of numeronyms for its input strings
- Create an executable to demonstrate
- Add unit tests

jonathanheaden commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,56 @@
+
+func fib(n int) {
+	if n > 20 || n < -20 {
+		panic("inconceivable")
```
You keep using that word. I do not think it means what you think it means.

camscale replied:
I can tell you that fib(9000000) is 1,880,889 digits long from using `math/big.Int`, so I think I could conceive of a number outside this range.

But more seriously, consider whether `panic` is appropriate here. `panic` in Go is used sparingly - in particular you should not make panic a part of a package's API. In this case you're not making a package but assuming the more general implementation, `fib` could easily be part of some `math` package. In that case, it should either return an error or print out the error the same way it prints out the results. Having the caller handle the panic is un-Go-like.

More details here: https://github.com/golang/go/wiki/PanicAndRecover#usage-in-a-package

undrewb replied:
Makes sense cam. Originally I had a print but changed it on a reviewers suggestion. The limit is arbitrary and I agree it could be bigger. I'll make it an error since I can see that panic is a bit over the top.

anzdaddy replied:
To be clear:
> Better to return an `error` or `panic`.

jonathanheaden reviewed (approved):
I like the use of recursion

runnerdave committed:
Implement function plus table tests

juliaogris reviewed (changes requested):
Please add review of colleague's PR to your own PR description.  
Please squash your 21 commits into one or two with `git rebase -i <HASH>` and `git push -f`.  
Have a read through the [Collected Code Review Comments](https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments)

Jake88 committed:
Merge branch 'master' into lab1

runnerdave committed:
Merge branch 'master' into lab01-fibonacci

runnerdave committed:
Merge branch 'master' into lab01-fibonacci

Commit:
Improve tests

Use testify
Get rid of tests helper function
Add test cases for empty structure

Commit:
Implement Bubble and Insertion

Commit:
Implement 02_bubble

nicholascross committed:
Implement IPAddr type

- Implement IPAddr type backed by an array of 4 bytes
- Implement IPAddr related function to print IP address
- Add simple application to demonstrate formatting
- Add unit tests to cover string formatting of IPAddr

nicholascross committed:
Implement IPAddr type

- Implement IPAddr type backed by an array of 4 bytes
- Implement IPAddr related function to print IP address
- Add simple application to demonstrate formatting
- Add unit tests to cover string formatting of IPAddr

runnerdave committed:
Merge branch 'master' into lab01-fibonacci

runnerdave committed:
Merge pull request #222 from runnerdave/lab01-fibonacci

Implement function plus table tests

runnerdave committed:
Implement Bubble sort function plus tests

runnerdave committed:
Implement Bubble sort function plus tests

Commit:
Lab 2 - Bubble sort (#132)

* Implement bubble sort
* Implement insertion sort
* Implement merge sort
* Improve unit-tests

odrij committed:
Lab1 Fibonacci and Negafibonacci number

odrij committed:
Tests for fibonacci function

ovlords committed:
LAB1: Implement function that produces Fibonacci sequence (#152)

* Implement function that produces Fibonacci sequence

Commit:
Modify fibonacci loop

Commit:
Fix miscellaneous according to CR

Fix overcomplicated logic in tests
Rename variable
Split tests into positive and negative

willshen24 committed:
Lab02 - Completed bubble and insertion sort with unit tests

jonathanheaden committed:
Merge branch 'master' into 01-fib

jonathanheaden committed:
Merge pull request #169 from jonathanheaden/01-fib

01 fib

pfannerj committed:
Merge branch 'master' into lab1

Commit:
Lab 3 - Letter frequency (#133)

* Refactor tests as table driven tests
* Improve test cases

pfannerj committed:
Merge branch 'master' into lab1

pfannerj committed:
Merge pull request #145 from pfannerj/lab1

Lab1  Fibonacci and bonus negafibonacci

willshen24 committed:
- Modifed sort functions to give immutability on original slice
- Updated code based on review comments

anzdaddy commented:
> The instruction also asks to provide not more than 2 commits (https://github.com/anz-bank/go-course/wiki/Lab-1:-Collected-Code-Review-Comments). Please consider rebasing.

Use common sense and consider whether the effort is a nett benefit.

n0npax committed:
Merge branch 'master' into lab05

n0npax committed:
Merge pull request #215 from n0npax/lab05

Complete lab05 - Stringer

lexual-anz committed:
testing main() function to get 100% test coverage

willshen8 committed:
Merge pull request #254 from willshen8/lab02

Lab02 - Bubble and insertion sort with unit tests

Commit:
Change test cases to be table driven and refine following PR review

lexual-anz committed:
Implement bubble and insertion sort

Commit:
Implement Numeronym

Implement Numeronym

runnerdave committed:
Implement feedback from PR review

runnerdave committed:
Implement feedback from PR review

runnerdave committed:
Merge branch 'master' into bubble-sort

Commit:
refactor(fib): Refactor fib function to use two variables instead of a map

Removed the map from fib function and replaced it with two integers to save memory.

Commit:
refactor(fib): Refactor fib function to use two variables instead of a map

Removed the map from fib function and replaced it with two integers to save memory.

Commit:
feat(fib): Add the ability to handle 0 as in input to fib

Supplying an input of 0 to fib will now print f0 (which is 0) in return.

Commit:
feat(fib): Add the ability to handle 0 as in input to fib

Supplying an input of 0 to fib will now print f0 (which is 0) in return.

Commit:
feat(fib): Add a check for integer overflow to fib function

Fib function now has a check to see if the input has caused a number overflow error and prints it out accordingly.
Added a few minor comment updates.

Commit:
feat(fib): Add a check for integer overflow to fib function

Fib function now has a check to see if the input has caused a number overflow error and prints it out accordingly.
Added a few minor comment updates.

Commit:
test(fib): Add tests for fib function

Added a test for fib(0) and a test to check whether an out of bounds / number overflow error has ocurred.

Commit:
test(fib): Add tests for fib function

Added a test for fib(0) and a test to check whether an out of bounds / number overflow error has ocurred.

Commit:
Merge branch 'lab1' of https://github.com/Jake88/go-course into lab1

Commit:
refactor(fib): Additional linting fix

Impatient me went and pushed the code before proof reading my comments. Additional commit to ensure they read correctly.

Commit:
refactor(fib): Additional linting fix

Impatient me went and pushed the code before proof reading my comments. Additional commit to ensure they read correctly.

n0npax committed:
Merge branch 'master' into lab03

lexual-anz committed:
Use table-driven testing. Better comment.

In responde to code review, adopted table-driven testing.

Also made sure function comment, had function as 1st
word in comment.

n0npax committed:
Merge pull request #209 from n0npax/lab03

Lab03 - Letter frequency

nicholascross committed:
Merge branch 'master' into lab3

nicholascross committed:
Merge pull request #201 from nicholascross/lab3

Implement sorted letter counting

lexual-anz committed:
Implement numeronym & numeronyms functions

lexual-anz committed:
Adopt subtests utiling t.Run()

undrewb commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,73 @@
+var out io.Writer = os.Stdout
+
+func main() {
+	err := fib(7)
```
fixed

undrewb commented on 01_fib/undrewb/main.go:
```diff
@@ -0,0 +1,72 @@
+var out io.Writer = os.Stdout
+
+func main() {
+	fib(7)
```
fixed

cuminandpaprika committed:
Add fibonacci sequence calculator

Add implementation of fibonacci calc

Add loop to print fibonacci sequences

Add handling of negative inputs 

Add fix to main to follow lab instructions

Fix linter Issues

Commit:
Build Lab4 Numeronym solution and test cases

jessodri committed:
Merge branch 'master' into master

jessodri committed:
Merge pull request #246 from jessodri/master

Lab1 Fibonacci and Negafibonacci number

nicholascross committed:
Merge branch 'master' into lab4

nicholascross committed:
Merge pull request #216 from nicholascross/lab4

Implement numeronyms

davidbroughsmyth committed:
Implement Fibonacci and Negafibonacci

Commit:
Add function to count letter frequency in string with tests

Commit:
Fix linter issue

davidbroughsmyth committed:
Merge pull request #313 from davidbroughsmyth/lab1

lab1 - Implement Fibonacci and Negafibonacci

willshen24 committed:
Implemented letters lab with test cases

- Solution check for existence first and then increment it in map
- Sort Letter is done by using sort.string()
- Test cases covers all different unicodes and empty inputs

Commit:
Implement Fibonacci Series

Implemented fibonacci for positive and negative numbers using for loop

Commit:
Adds Fibbonacci and Tests

Commit:
Fix lint error

Commit:
Fix lint error

willshen24 committed:
Changes added after PR review:
-use testify testing library
-removed branching logic for letter mapping
-use better slice declaration for performance tuning

n0npax committed:
Complete lab06

Implement CRUD for pet structure
Implement Sync and mem storer
Implement tests

lexual committed:
Merge branch 'master' into lab02

lexual committed:
Merge pull request #261 from lexual/lab02

Implement bubble and insertion sort

willshen8 committed:
Merge branch 'master' into lab03

willshen8 committed:
Merge pull request #333 from willshen8/lab03

Lab03 - Letters Frequency

willshen24 committed:
-Completing Lab05 - Stringer with test cases
-Adding changes after PR reviews

camscale committed:
travis: Change golangci-lint download script

Travis is currently failing when running the script to download
golangci-lint, with little output to help show what's going wrong:

$ curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $(go env GOPATH)/bin v1.16.0
golangci/golangci-lint info checking GitHub for tag 'v1.16.0'
golangci/golangci-lint info found version: 1.16.0 for v1.16.0/linux/amd64
The command "curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $(go env GOPATH)/bin v1.16.0" failed and exited with 1 during .

Pull the script directly from golangci-lint's github repo instead of
using install.goreleaser.com, which seems to be breaking the download
script.

lexual committed:
Merge branch 'master' into lab1

camscale committed:
Merge pull request #372 from anz-bank/enable-travis-ci-download-debug

travis: Change golangci-lint download script

lexual committed:
Merge branch 'master' into lab1

lexual committed:
Merge pull request #148 from lexual/lab1

Lab1

Commit:
Implement Fibonaci and NegaFibinacci (#179)

* Implement recursion based function to return a slice of
  Fibonacci numbers starting at 0
* Implement functions to present fibonacci as a string / stdout

n0npax committed:
Impelent CR fixes

Improved error messages
Optimalize tests
Improve syntax
Add comments to exported types

brvishnu92 committed:
Merge branch 'master' into lab1

n0npax committed:
Merge branch 'master' into lab06

n0npax committed:
Merge pull request #358 from n0npax/lab06

Complete lab06

brvishnu92 committed:
Merge branch 'master' into lab1

brvishnu92 committed:
Merge pull request #174 from brvishnu92/lab1

Implement Lab1

pfannerj committed:
Build Lab5 Stringer solution and test cases

alfredxiao committed:
initial commit

alfredxiao committed:
Implement fibonacci series, tested edge cases and pos/negatives

nicholascross committed:
Merge branch 'master' into lab5

nicholascross committed:
Merge pull request #238 from nicholascross/lab5

Implement IPAddr type

Jake88 committed:
Merge branch 'master' into lab1

Jake88 committed:
Merge pull request #195 from Jake88/lab1

Lab1 - Fibonacci function

pentaphobe committed:
Add @pentaphobe to CODEOWNERS

One more limb to the Voltron that is course reviewers

pentaphobe committed:
Add @pentaphobe to CODEOWNERS

One more limb to the Voltron that is course reviewers

pentaphobe committed:
Merge pull request #406 from pentaphobe/master

Add @pentaphobe to CODEOWNERS

Commit:
Raise error when big number is being requested while not being supported!

runnerdave committed:
Merge branch 'master' into bubble-sort

runnerdave committed:
Merge pull request #243 from runnerdave/bubble-sort

Implement Bubble sort function plus tests

Commit:
Test panic situation where number out of range

willshen8 committed:
Merge branch 'master' into lab05

willshen8 committed:
Merge pull request #369 from willshen8/lab05

Lab05 - Stringers

brnbke committed:
Lab3 adding letters and sortLetters Functions (#391)

n0npax committed:
complete lab08

Refactor lab07 to fit project layout
Implement new tests

pfannerj committed:
Merge branch 'master' into lab4

pfannerj committed:
Merge pull request #376 from pfannerj/lab4

Lab4 Numeronym

Commit:
Stringer (#363)

* type IPAddr represents a IPv4 address and implements Stringer
  unspecified bytes are defaulted to 0

* main() prints the localhost IPv4 address followed by newline

saurabhvyas9 committed:
Merge pull request #239 from saurabhvyas9/lab2

Implement 02_bubble Lab2

saurabhvyas9 committed:
01_fib Lab1 commit for review - saurabhvyas9. (#225)

Implement 01_fib Lab1

juliaogris commented on 01_fib/kasunfdo/main.go:
```diff
@@ -0,0 +1,36 @@
+package main
+
```
This file still doesn't belong in this PR ;) (same below)

cuminandpaprika committed:
Fix PR Issues

Refactor to use int64 to handle larger range of inputs

Remove duplicated calcFib logic and tests


Refactor out handling of zero case earlier



Refactor to use bitwise and for evenness 

Remove binary file

Commit:
Simplify the series calculation logic

alfredxiao committed:
Merge branch 'master' into lab1-alfredxiao

Commit:
Implement Bubble/Insertion/Quick Sort (#242)

* New function bubble([]int) []int returns sorted copy
  of passed in slice using bubble sort
* New function insertion([]int) []int returns sorted copy
  of passed in slice using insertion sort
* New function quick([]int) []int returns sorted copy
  of passed in slice using quick sort
  Uses Lomuto partition scheme

alfredxiao committed:
Merge branch 'master' into lab1-alfredxiao

pfannerj committed:
Merge branch 'master' into lab5

pfannerj committed:
Merge pull request #382 from pfannerj/lab5

Lab5 Stringer

pfannerj committed:
Merge branch 'master' into lab2

pfannerj committed:
Merge pull request #252 from pfannerj/lab2

Lab2 Bubble sort and bonus Insertion sort

brnbke committed:
Complete Lab 1 (#212)

Building function to create Fibonacci and Negafibonacci numbers

jessodri committed:
Merge branch 'master' into lab3

sirishavadrevu committed:
Implemented Fibonacci sequence and tests Lab1

jessodri committed:
Merge pull request #323 from jessodri/lab3

Lab 3 - Letter Frequency

saurabhvyas9 committed:
Implement Lab5 05_Stringer (#466)

raghunadhpokkalath committed:
Merge pull request #343 from raghunadhpokkalath/lab01

Lab01

brvishnu92 committed:
Merge branch 'master' into lab2

brvishnu92 committed:
Merge pull request #219 from brvishnu92/lab2

Implement Bubble and Insertion - Lab 2

alfredxiao committed:
Merge branch 'master' into lab1-alfredxiao

alfredxiao committed:
Merge pull request #386 from alfredxiao/lab1-alfredxiao

Lab1 fibonacci - alfredxiao

Commit:
modules: Change module name to go-course from go-training

When the repository was renamed from go-training to go-course, the
go.mod file was not updated to reflect the new name.

Issues: #479

camscale committed:
Merge pull request #480 from camh-anz/fix-go-module-name

modules: Change module name to go-course from go-training

juliaogris committed:
Update lab 08_project REAMDE.md

Remove `go.mod` and `go.sum` from required lab_08 directory layout because
having `go.mod`  file in root and sub-directory causes tests not to be run in
sub-directory.

Remove `vendor` directory from required lab_08 directory layout as it is not
going to be used. The vendor directory is intended for a copy of all
dependencies and useful in locked-down environments where some dependencies
might not be accessible by other means. However, using them on a per lab basis
is not useful and will blow out the repo size.

Remove USERNAME from `pkg` and `cmd`  sub-directory names. Separating by
username has been lifted up to `08_project/USERNAME`. This has been missed
in previous clean-up efforts.

Re-generate `labs.md` with:

	for i in */README.md; do cat "$i"; echo; done > labs.md

Commit:
Add leveldb to go.mod and go.sum

n0npax committed:
Merge pull request #481 from anz-bank/feature/add-lvldb-to-go-mod

Add leveldb to go.mod and go.sum

juliaogris committed:
Merge pull request #482 from juliaogris/fix-lab8-spec

Update lab 08_project REAMDE.md

Commit:
Preallocate slice with known size

n0npax committed:
Implement CR fixes

n0npax committed:
Merge pull request #422 from n0npax/lab08

Complete lab08

lexual committed:
Merge branch 'master' into lab04

lexual committed:
Merge pull request #274 from lexual/lab04

Implement numeronym & numeronyms functions

joel00wood committed:
Implemented Numeronyms (#390)

Created `numeronym()` and `numeronyms()` functions. `numeronym()` takes a string as input and returns a string with the first and last character wrapped around it's length -2. `numeronyms()` takes a variadic string slice as input and returns a slice of those strings converted to numeronyms.

jonathanheaden committed:
Implement bubble sort and add tests (#206)

Handle the case for zero element slice by returning the slice unchanged
Add tests for zero element slice & all elements the same (4 nines)

n0npax committed:
Merge branch 'master' into lab2

n0npax committed:
Merge pull request #182 from n0npax/lab2

Lab 2 - Sorting

cuminandpaprika committed:
Merge branch 'master' into lab01

cuminandpaprika committed:
Merge pull request #281 from cuminandpaprika/lab01

Lab1

sirigithub committed:
Merge branch 'master' into sirigithublab1

Commit:
Lab 4 - Numeronym (#134)

* Implement numeronym
* Support unicode characters, improve unit tests
* Refactor according to the review

sirigithub committed:
Merge branch 'master' into sirigithublab1

sirigithub committed:
Merge pull request #467 from sirigithub/sirigithublab1

Implemented Fibonacci sequence and tests Lab1

mohitnag committed:
Merge pull request #339 from mohitnag/lab1

Adds Fibbonacci and Tests

Commit:
Letter Frequency (#347)

* letters(string) map[rune]int
  returns a map of the frequency of runes

* sortLetters(map[rune]int) [] string
  returns a slice of rune:frequency pairs formatted as string
  and ordered by descending frequency

* main()
  prints an ordered list of rune:freuency pairs

willshen8 committed:
Lab04 - Numeronyms (#337)

* Implemented numeronyms function using regex with full coverage test cases

* Fixing code after PR review

Commit:
Allow duplicated code in golangci conf

Turn off duplicate checking because labs 06 - 11 can have the same files,
because they progressively build on the previous lab.
With duplicate checking enabled, travis checks fail.

n0npax committed:
Merge branch 'master' into bugfix/allow-dupl-in-golangci

n0npax committed:
Merge pull request #497 from anz-bank/bugfix/allow-dupl-in-golangci

Allow duplicated code in golangci conf

joel00wood committed:
Created Fibonacci sequencer (#227)

Built fib(n int) as 3 functions. The base function `fibCalc`
calculates the Nth fibonacci value. The wrapper function `fibN` enables
support for Negafibonacci calculation. `fib` itself just prints 0 to
(+|-)n, on each line.

Borrowed output capture function from elsewhere to
enable testing of `fib` and `main` that do not return values.

saurabhvyas9 committed:
Implement 03_letters (#433)

Commit:
Numeronyms (#355)

Implement numeronym requirements in a UTF8 aware way.

* numeronym(string) string returns the numeronym for a string

* numeronyms(string...) []string returns a slice of numeronyms from input arguments

* main() prints numeronym slice for 3 commonly abreviated words

brvishnu92 committed:
Merge branch 'master' into lab4

brvishnu92 committed:
Merge pull request #265 from brvishnu92/lab4

Implement Lab4 -  Numeronym

juliaogris reviewed (approved):
Code passes, commits are neat. LGTM
As mentioned in person, here's our canonical solution for comparison:

```go
func fib(n int) {
	sign := 1
	if n < 0 {
		sign = -1
	}

	n1, n2 := 0, 1
	for j := 0; j < n*sign; j++ {
		n1, n2 = n2, n1+sign*n2
		fmt.Fprintln(out, n1)
	}
}
```
